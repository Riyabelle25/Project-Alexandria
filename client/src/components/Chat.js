import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import {
  SendRounded,
} from '@material-ui/icons';
import firebase from 'firebase';
import '../styles/Chat.css';

import ChatHeader from './ChatHeader';
import Message from './Message';

import db from '../app/firebase';
import { selectUser } from '../slice/userSlice';
import { selectChannelId, selectChannelName } from '../slice/appSlice';
import { useParams } from 'react-router-dom';

function Chat() {
  const user = useSelector(selectUser);
  const channelId = useSelector(selectChannelId);
  const channelName = useSelector(selectChannelName);
  const [messageInput, setMessageInput] = useState('');
  const [messages, setMessages] = useState([]);
  const roomName = useParams().name;

  useEffect(() => {
    if (channelId) {
      db.collection("rooms").doc(roomName).collection('channels')
        .doc(channelId)
        .collection('messages')
        .orderBy('timestamp', 'asc')
        .onSnapshot((snapshot) => {
          setMessages(snapshot.docs.map((doc) => doc.data()));
        });
    }
  }, [channelId]);

  const sendMessage = (e) => {
    e.preventDefault();
    db.collection("rooms").doc(roomName).collection('channels').doc(channelId).collection('messages').add({
      timestamp: firebase.firestore.FieldValue.serverTimestamp(),
      message: messageInput,
      user,
    });
    setMessageInput('');
  };

  return (
    <div className="chat">
      <ChatHeader channelName={channelName} />
      <div className="chat__messages">
        {messages.map((message, index) => (
          <Message
            key={index}
            message={message.message}
            timestamp={message.timestamp}
            user={message.user}
          />
        ))}
      </div>
      <div className="chat__input">
        <form>
          <input
            disabled={!channelId}
            onChange={(e) => setMessageInput(e.target.value)}
            placeholder={
              channelName
                ? `Message #${channelName}`
                : 'Please select a channel first!'
            }
            type="text"
            value={messageInput}/>    
            <button
            className="chat__inputButton_hidden"
            onClick={sendMessage}
            type="submit"
          >
            Send Message
          </button>  
        </form>
        <SendRounded className="chat__inputButton" onClick={sendMessage} type="submit"/>    
      </div>
    </div>
  );
}

export default Chat;
