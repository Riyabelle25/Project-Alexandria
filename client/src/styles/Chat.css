.chat {
  display: flex;
  flex-direction: column;
  flex: 0.78;
  background-color: rgb(13, 1, 27);
  height: 100vh;
}

.chat__messages {
  overflow-y: scroll;
  flex: 1;
}
  /* width */
  .chat__messages::-webkit-scrollbar {
    width: 8px;
    height: 75vh;
  }
  
  /* Track */
  .chat__messages::-webkit-scrollbar-track {
    background: rgb(13, 1, 27);
  }
  
  /* Handle */
  .chat__messages::-webkit-scrollbar-thumb {
    background: rgb(48, 41, 54);
  }
  
  /* Handle on hover */
  .chat__messages::-webkit-scrollbar-thumb:hover {
    background: #555;
  }
.chat__input {
  display: flex;
  justify-content: space-around;
  align-items: center;
  padding: 15px;
  color: rgb(173, 187, 181);
  border-radius: 5px;
  margin: 20px;
  border-top: 1px solid rgb(48, 41, 54);
  background-color: rgb(48, 41, 54);
}

.chat__input > form > input {
  padding: 15px;
  background: transparent;
  border: none;
  outline-width: 0;
  color: rgb(13, 1, 27);  
  font-size: large;
  width: 100%;
}

.chat__input > form {
  flex: 1;
}


.chat__inputIcons > .MuiSvgIcon-root {
  padding: 5px;
}
.chat__inputButton_hidden{
  display:none;

}

