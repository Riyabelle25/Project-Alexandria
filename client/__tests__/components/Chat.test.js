import Chat from "../../src/components/Chat";
import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import toJson from "enzyme-to-json";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store"; // Smart components

configure({ adapter: new Adapter() });

describe("<Chat />", () => {
  const initialState = { user: null };
  const mockStore = configureStore();
  let store;
  describe("render()", () => {
    test("renders the component", () => {
      store = mockStore(initialState);
      const wrapper = shallow(
        <Provider store={store}>
          <Chat />
        </Provider>
      );
      const component = wrapper.dive();

      expect(toJson(component)).toMatchSnapshot();
    });
  });
});
