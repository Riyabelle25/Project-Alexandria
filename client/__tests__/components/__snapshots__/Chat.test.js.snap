// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> render() renders the component 1`] = `<Chat />`;
